src/veins/modules/mobility/traci/subscriptionManagement/SubscriptionManagerBase.h

```diff
@@ -0,0 +1,162 @@
+//
+// You should have received a copy of the GNU General Public License
+// along with this program; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+//
+
+#ifndef SRC_VEINS_MODULES_MOBILITY_TRACI_SUBSCRIPTIONMANAGEMENT_SUBSCRIPTIONMANAGERBASE_H_
+#define SRC_VEINS_MODULES_MOBILITY_TRACI_SUBSCRIPTIONMANAGEMENT_SUBSCRIPTIONMANAGERBASE_H_
+
+#include <string>
+#include <set>
+#include <memory>
+
+#include "veins/modules/mobility/traci/TraCIBuffer.h"
+#include "veins/modules/mobility/traci/TraCICommandInterface.h"
+#include "veins/modules/mobility/traci/TraCIConnection.h"
+
+
+namespace veins {
+namespace TraCISubscriptionManagement {
+
+/**
+ * @class SubscriptionManagerBase
+ *
+ * This class provides some common functionality for a subscription manager. It stores
+ * TraCIConnection, TraCICommandInterface and a set containing subscribed ids.
+ *
+ * It also enforces you to provide an update() method in your subscription manager.
+ *
+ * @author Nico Daßler <dassler@hm.edu>
+ */
+class SubscriptionManagerBase {
+
+public:
+    /**
+     * Constructor.
+     */
+    SubscriptionManagerBase();
+
+    /**
+     * Default destructor.
+     */
+    virtual ~SubscriptionManagerBase() = default;
+
+    /**
+     * Each subscription manager has to implement this method. The given buffer
+     * is assumed to have the first few bytes removed. These bytes represent
+     * the first few fields of an answer to a simulation step request or the
+     * request to subscribe variable request.
+     * The fields that should be removed are:
+     *      - the number of subscription results
+     *      - the response command length
+     *      - the response command length extended
+     *      - the id of the response commmand
+     *
+     * This method is expected to change the results to api methods of this manager.
+     *
+     * @param buffer the TraCIBuffer containing the subscription response.
+     *
+     * @return if there was an id_list was received in the buffer (not all subscription manager will use
+     * this. the result is undefined for them).
+     */
+    virtual bool update(TraCIBuffer& buffer) = 0;
+
+    /**
+     * This initializes this manager with the necessary connection and command interface
+     * to perform requests to the TraCI server.
+     *
+     * Important: Always do this before using a subscription manager.
+     *
+     * Note: When overriding the this method from a child class make sure to call the base
+     * class implementation to set mConnection and mCommandInterface variables.
+     *
+     * @param connection to access traci server (for example: to perform subscriptions).
+     * @param commandInterface to access traci (for example: direct request to get vehicle id_list)
+     */
+    virtual void initialize(std::shared_ptr<TraCIConnection> connection, std::shared_ptr<TraCICommandInterface> commandInterface);
```

Comment:
When I change the value of a shared_ptr does that change become visible in all classes holding that pointer?

Reply:
Yes, the object behind the `shared_ptr` is the same for all if its instances. In existing code however, we used regular pointers to indicate that they are non-owning. The change to `shared_ptr` seems like it does not reflect the semantics properly - the connection's owner is the `TraCIScenarioManager`, and also it is incompatible with the current code base where, for that reason, we use a `unique_ptr`. Is there a particular reason you opted to change this to `shared_ptr`s?

A less invasive integration with the `TraCIScenarioManager` would be to use its `traci_initialized` signal, this way you can avoid integrating it deeply within it.
However, since the subscription managers it might be worth putting some thought into what constitutes TraCI being initialized: Does this require that the subscriptions are being set up? If so, we might want to add TraCI pre-, post-, and actual-initialization signals.

Reply:
Hm. I used a `shared_ptr` because I thought (might be wrong) that it is not possible to simply pass around a `unique_ptr`. (I'm not sure but I think my expectation got proven correct when I got a compilation error. Mabe used it wrong...).

The reason why I need to pass these pointers is because the different managers need access to them to perform subscriptions.

In my opinion the best solution would be a "global" object that is instantiated in the manager and then automatically shared with all managers. Then there will be a call to a method (something like `setUpSubscriptions`) that all managers implement which set ups the basic subscriptions.

I can't really comment on the second paragraph since i do not understand what it means "to use its `traci_initialized` signal". (I'm sorry I'm still a beginner with omnet++).

Reply:
I think the easiest and cleanest way to solve (or get around) this problem is to only construct all subscription managers once the `TraCIConnection` is established (and subsequently, `TraCICommandInterface` is available, too). Then, we can get rid of this whole `initialize` method for the subscription managers and they cannot get into an uninitialized state.

Whether to use a raw or shared pointer (or even a reference, which is possible if we pass it in at construction, and thus would be my suggestion) is still open, but only a minor issue.

So I would do the following:
- Once TraCI is initialized (which can be detected by the `TraCIScenarioManager::traci_initialized` simulation signal (see [the omnet++ manual about signals](https://doc.omnetpp.org/omnetpp/manual/#sec:simple-modules:signals) and [some info about using those signals in veins](https://github.com/sommer/veins/blob/master/doc/manual/simsignal_management.md)):
- The module owning the `SubscriptionManager` (probably the `TraCIScenarioManager`) constructs the `SubscriptionManager`, passing in the `TraCIConnection` and `TraCICommandInterface` (by reference or pointer). As this happens after the construction of the owning module, it should be constructed into a `std::unique_ptr`.
- The `SubscriptionManager` in turn constructs the individual subscription managers, handing on `TraCIConnection` and `TraCICommandInterface` references/pointers. The individual subscription managers are just constructed as data members (no pointers) and valid/initialized immediately after construction.
- The whole system is ready and can start processing subscriptions.

Reply:
@dbuse I like the design you describe.
@truenicfel You're right, you cannot create copies of a `unique_ptr` around as it has ownership semantics. However, you can access the underlying pointer and treat it as a 'weak pointer', i.e., a reference that will not keep the object alive.

Reply:
So here is what I did now:

- `unique_ptr<ExecutiveSubscriptionManager> subscriptionManager;` as class variable in `TraCIScenarioManager` which is initialized in the `init_traci()` method of the scenario manager
- two shared ptrs for command interface and connection in `TraCIScenarioManager` 
- the two shared ptrs will be passed (by reference) into the constructor of everything related to subscriptions

Please note that I will push the changes to `TraCIScenarioManager` when we are finished with every file in directory `.../subscriptionManagement` (see comment below).